src/mmap.rs

```diff
@@ -0,0 +1,583 @@
+                -1,
+                0,
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(Self {
+            addr: addr as *mut u8,
+            size,
+        })
+    }
+
+    /// Maps the `size` bytes starting at `offset` bytes of the given `fd`.
+    ///
+    /// # Arguments
+    /// * `fd` - File descriptor to mmap from.
+    /// * `size` - Size of memory region in bytes.
+    /// * `offset` - Offset in bytes from the beginning of `fd` to start the mmap.
+    pub fn from_fd(fd: &AsRawFd, size: usize, offset: libc::off_t) -> io::Result<Self> {
+        // This is safe because we are creating a mapping in a place not already used by any other
+        // area in this process.
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+        Ok(Self {
+            addr: addr as *mut u8,
+            size,
+        })
+    }
+
+    /// Returns a pointer to the beginning of the memory region.  Should only be
+    /// used for passing this region to ioctls for setting guest memory.
+    pub fn as_ptr(&self) -> *mut u8 {
+        self.addr
+    }
+
+    unsafe fn as_slice(&self) -> &[u8] {
+        // This is safe because we mapped the area at addr ourselves, so this slice will not
+        // overflow. However, it is possible to alias.
+        std::slice::from_raw_parts(self.addr, self.size)
+    }
+
+    // safe because it's expected interior mutability
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn as_mut_slice(&self) -> &mut [u8] {
+        // This is safe because we mapped the area at addr ourselves, so this slice will not
+        // overflow. However, it is possible to alias.
+        std::slice::from_raw_parts_mut(self.addr, self.size)
+    }
+}
+
+impl VolatileMemory for MmapRegion {
+    fn len(&self) -> usize {
+        self.size
+    }
+
+    fn get_slice(&self, offset: usize, count: usize) -> volatile_memory::Result<VolatileSlice> {
+        let end = calc_offset(offset, count)?;
+        if end > self.size {
+            return Err(volatile_memory::Error::OutOfBounds { addr: end });
+        }
+
+        // Safe because we checked that offset + count was within our range and we only ever hand
+        // out volatile accessors.
+        Ok(unsafe { VolatileSlice::new((self.addr as usize + offset) as *mut _, count) })
+    }
+}
+
+impl Drop for MmapRegion {
+    fn drop(&mut self) {
+        // This is safe because we mmap the area at addr ourselves, and nobody
+        // else is holding a reference to it.
+        unsafe {
+            libc::munmap(self.addr as *mut libc::c_void, self.size);
+        }
```

Comment:
A few comments regarding this behavior.

First, this assumes that part of the memory security is enforced through the code, instead of being enforced by the language itself. The assumption that nobody else is holding a reference is very hard to enforce...

The way `mmap()` behaves, we can clearly expect that `mmap()` would be called multiple times on the same address with different ranges, without having to call into `munmap()`. Calling `munmap()` would prevent us from atomically change the mapping at a specific fixed address.
This is a very specific case that I ran into while porting virtio-fs for Firecracker (DAX case), since the memory mapping that happens on behalf of the libfuse daemon is performed on top of the existing mapping holding the whole shared region.

Problem is, by having this `Drop` behavior, this forces the virtio-fs device to maintain a list of `MemoryMapping` structures if we don't want to see the mapping disappear. And even by maintaining a list, what happens when someone wants to `mmap()` on top of an address that is already tracked?

Ok let's try to take an example:
1. Map 16MB of anonymous pages at address 0x0
2. Map 4MB of pages from a file descriptor at address 0x200000 (2MB)
3. At this point, you have two entries in the list of `MemoryMapping`
4. Unmap 16MB of pages at address 0x0
5. At this point you only have one entry corresponding to the second mapping
6. Trying to access the page at address 0x200000 will generate a segfault

What I'm trying to say is that by enforcing the unmap from the code itself, we need deterministic behavior. In the case of virtio-fs, the hypervisor is doing actions (`mmap()` basically) on behalf of the vhost-user backend, that's why it seems pretty hard to maintain proper state every time a `mmap()` happens.

My question is, do we want to keep this behavior for any memory mapping, by maintaining a complex list of descriptors, or do we simply want to create another type of `MmapRegion`, like `MmapRegionRaw` that would not include any `Drop` behavior?

Reply:
@sboeuf Actually this now really a `MmapRegionRaw`. This is a special type of region, where you want to free mapping to the mapped contents but not give us the virtual memory reservation.

To be able to do this atomically you cannot do an unmap, but you need to do a mmap of anonymous memory at the same virtual address with the same same size. Hence the VMM gets to hold on to the virtual address reservation.

The only assumption here is that there is an overarching region that covers all of these *special" region that will outlive all of these sub regions and that will be responsible for the actual mmap.

Reply:
@mcastelino you're right! Having a special type handling the `Drop` with `mmap(MAP_ANONYMOUS)` instead of `munmap()` looks like the only clean way to address this issue.

@jiangliu @bonzini WDYT?

Reply:
I didn't know about `std::mem::forget`, but this could be a potential solution to the problem too.
That being said, @mcastelino suggestion would be the cleanest approach IMO since we would make sure the memory region is reserved without leaving the previous mapping.

This can be part of a follow up PR, I don't mind, but I want to make sure we discuss and decide how we want to solve this asap, as this may influence the current code.

Reply:
It's a special requirement to support virtio-fs, and we haven't considered such a use case before.

Let's first talk about the requirements of the virtio-fs driver. Per my understanding, to support virtio-fs devices, an MMIO window is created in guest's physical address space, and the virtiofsd dynamically  map/remap contents from files into the MMIO window. Is that correct?

If it's so, how about separating the virtio-fs logic from the vm-memory crate? We could do it in this way:
1) create a GuestRegionMmap object to allocate the virtual address and optionally anonymous memory for the MMIO window.
2) introduce a new VirtioFsManager to manage the virtual address allocated in step 1  as a pool and provides interfaces to
     a) allocate virtual address from the pool to map file content into the MMIO window.
     b) free virtual addresses into the pool when a file is deleted/closed.
     c) reclaim virtual address by LRU when the pool runs out of virtual address.
     d) map/remap file content to specific virtual address with mmap(MAP_FIXED).
3) the VirtioFsManager object out-lives the GuestRegionMmap object for the MMIO window.
By this way, we don't need to change the vm-memory crate and just introduce a new manager for virtio-fs.

As a bonus, we may make a minor improvement to the vm-memory crate to support mmap(PROT_NONE) in step 1, so that page fault happens when guest accesses a virtual address in the MMIO window which hasn't been assoicated with any file content yet.

Reply:
@jiangliu I am not sure if this requirement is limited to virtio-fs. Will we have a need for something similar when managing the PCI Hole and BAR's within the PCI Hole. @bonzini may have a better idea. Will the reqions corresponding to the BARs be allocated on the fly. If the regions corresponding to the BARs are allocated on the fly, then the current implementation of `GuestMemoryMmap` should probably need to support adding regions.

Reply:
@jiangliu 
>Let's first talk about the requirements of the virtio-fs driver. Per my understanding, to support virtio-fs devices, an MMIO window is created in guest's physical address space, and the virtiofsd dynamically map/remap contents from files into the MMIO window. Is that correct?

Yes this is correct.

Introducing a `VirtioFsManager` seems like a lot to address a problem that is more generic as we could end up doing multiple `mmap` in any situation that requires submappings withing an existing primary mapping.

With your solution, you would create a new entry holding onto an `MmapRegion` every time a new submapping happens. And at the end, when all those entries go out of scope, they get dropped with `Drop`, which means a bunch of `munmap()` are being called. This works but seems more complex than the solution proposed by @mcastelino.

Eventually, I think we would want to be able to define that a `MmapRegion` holds a list of submapping that could be implementing a different `Drop` (doing an `mmap` instead).

Last thing, the mapping could be handled by a manager as you proposed, but if we end up going down this path, we should have this manager being defined as a trait that can be reused by any device. We cannot expect every device to handle this use case in its own way.